### ISSUE TITLE
bug: SNS with SQS and FIFO does only work for the first message

### ISSUE THREAD

**Issue by dandsafe** [localstack/localstack#10088]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I create a SNS and SQS, both with FIFO.
I would like to set my deduplicationId by myself.

In this case the first message arrives, the second message never arrives.

### Expected Behavior

I create a SNS and SQS, both with FIFO.
I would like to set my deduplicationId by myself.

Alle messages sent to SNS should arrive the SQS Queue

### How are you starting LocalStack?

localstack-cli

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

localstack start

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)
```
# Create SNS and SQS

awslocal sns create-topic --name "mysns.fifo" --attributes "{\"FifoTopic\":\"true\",\"ContentBasedDeduplication\":\"false\"}"

awslocal sqs create-queue --queue-name mysqs.fifo --attributes "{\"FifoQueue\":\"true\",\"ContentBasedDeduplication\":\"false\"}"

awslocal sns subscribe --topic-arn "arn:aws:sns:us-east-1:000000000000:mysns.fifo" --protocol sqs --notification-endpoint "arn:aws:sqs:us-east-1:000000000000:mysqs.fifo"

## send 1. message

awslocal sns publish --topic-arn "arn:aws:sns:us-east-1:000000000000:mysns.fifo" --message-group-id "mygroupid" --message-deduplication-id "superunique1" --message "supermessage"

awslocal sqs receive-message --queue-url http://localhost:4566/000000000000/mysqs.fifo

awslocal sqs delete-message --queue-url http://sqs.us-east-1.localhost.localstack.cloud:4566/000000000000/mysqs.fifo --receipt-handle "NjA4YzEzMzMtNDlkZS00NDI4LWFhM2EtNTY2M2I0MjU5Y2Y3IGFybjphd3M6c3FzOnVzLWVhc3QtMTowMDAwMDAwMDAwMDA6a2FtbWVyX2dldm9fZWluZ2VnYW5nZW4uZmlmbyA2ZTlmZGUyZS1kNzlmLTQ5YjgtOTQ5Yi00NWZiNWM0NTM4MGMgMTcwNTU4NDUzNS43ODI5NTUy"

## send 2. message with another deduplication-id, but same message
awslocal sns publish --topic-arn "arn:aws:sns:us-east-1:000000000000:mysns.fifo" --message-group-id "mygroupid" --message-deduplication-id "superunique2" --message "supermessage"

awslocal sqs receive-message --queue-url http://localhost:4566/000000000000/mysqs.fifo

# No message arrives here.

```

### Environment

```markdown
- OS:Windows 11
- LocalStack: 3.0.2
```

**Comment by bentsku** [comment]
Hi @dandsafe and thanks for your report!

I could not reproduce the issue, and I wonder if this is related to the message visibility or something similar.

I've ran the following:
```bash
$ awslocal sqs create-queue --queue-name mysqs.fifo --attributes "{\"FifoQueue\":\"true\",\"ContentBasedDeduplication\":\"false\"}"
{
    "QueueUrl": "http://localhost.localstack.cloud:4566/queue/us-east-1/000000000000/mysqs.fifo"
}
$ awslocal sns create-topic --name "mysns.fifo" --attributes "{\"FifoTopic\":\"true\",\"ContentBasedDeduplication\":\"false\"}"
{
    "TopicArn": "arn:aws:sns:us-east-1:000000000000:mysns.fifo"
}
$ awslocal sns subscribe --topic-arn "arn:aws:sns:us-east-1:000000000000:mysns.fifo" --protocol sqs --notification-endpoint "arn:aws:sqs:us-east-1:000000000000:mysqs.fifo"
{
    "SubscriptionArn": "arn:aws:sns:us-east-1:000000000000:mysns.fifo:a04a4c8a-1518-429a-9b0d-eb3f9ec5b8e1"
}
$ awslocal sns publish --topic-arn "arn:aws:sns:us-east-1:000000000000:mysns.fifo" --message-group-id "mygroupid" --message-deduplication-id "superunique1" --message "supermessage"
{
    "MessageId": "72677fac-ecfe-4548-a7f6-cd4984ce5763",
    "SequenceNumber": "14656133232500670464"
}
$ awslocal sqs receive-message --queue-url http://localhost.localstack.cloud:4566/queue/us-east-1/000000000000/mysqs.fifo --attribute-names All
{
    "Messages": [
        {
            "MessageId": "820b1acf-e3c3-4753-bc7b-f289fc8a2e04",
            "ReceiptHandle": "NzQyNjhhMTAtMTBiNS00NThjLThmMmMtMDA3MGJiZTI0ZjJiIGFybjphd3M6c3FzOnVzLWVhc3QtMTowMDAwMDAwMDAwMDA6bXlzcXMuZmlmbyA4MjBiMWFjZi1lM2MzLTQ3NTMtYmM3Yi1mMjg5ZmM4YTJlMDQgMTcwNjE5ODM3MS4yODQ5NDM=",
            "MD5OfBody": "38a7357e8f86c8dd3ee1cde4fde775c2",
            "Body": "{\"Type\": \"Notification\", \"MessageId\": \"72677fac-ecfe-4548-a7f6-cd4984ce5763\", \"TopicArn\": \"arn:aws:sns:us-east-1:000000000000:mysns.fifo\", \"Message\": \"supermessage\", \"Timestamp\": \"2024-01-25T15:59:27.359Z\", \"UnsubscribeURL\": \"http://localhost.localstack.cloud:4566/?Action=Unsubscribe&SubscriptionArn=arn:aws:sns:us-east-1:000000000000:mysns.fifo:a04a4c8a-1518-429a-9b0d-eb3f9ec5b8e1\", \"SequenceNumber\": \"14656133232500670464\"}",
            "Attributes": {
                "SenderId": "000000000000",
                "SentTimestamp": "1706198367389",
                "MessageGroupId": "mygroupid",
                "MessageDeduplicationId": "superunique1",
                "SequenceNumber": "14656132373507211264",
                "ApproximateReceiveCount": "1",
                "ApproximateFirstReceiveTimestamp": "1706198371284"
            }
        }
    ]
}
$ awslocal sqs delete-message --queue-url http://localhost.localstack.cloud:4566/queue/us-east-1/000000000000/mysqs.fifo --receipt-handle NzQyNjhhMTAtMTBiNS00NThjLThmMmMtMDA3MGJiZTI0ZjJiIGFybjphd3M6c3FzOnVzLWVhc3QtMTowMDAwMDAwMDAwMDA6bXlzcXMuZmlmbyA4MjBiMWFjZi1lM2MzLTQ3NTMtYmM3Yi1mMjg5ZmM4YTJlMDQgMTcwNjE5ODM3MS4yODQ5NDM=
(.venv) benjaminsimon@Benjamins-MacBook-Pro-2 tmp % awslocal sns publish --topic-arn "arn:aws:sns:us-east-1:000000000000:mysns.fifo" --message-group-id "mygroupid" --message-deduplication-id "superunique2" --message "supermessage"
{
    "MessageId": "caab9f3c-d4ef-4f1e-ba7c-6ba503680f1a",
    "SequenceNumber": "14656133232500670465"
}
$ awslocal sqs receive-message --queue-url http://localhost.localstack.cloud:4566/queue/us-east-1/000000000000/mysqs.fifo --attribute-names All
{
    "Messages": [
        {
            "MessageId": "f735f57f-af08-491b-a189-4c6bc1adbcf3",
            "ReceiptHandle": "YTI1OGE2M2QtZWVjMC00MTljLWFmNmEtYjVlOWY2NmE3NGQ1IGFybjphd3M6c3FzOnVzLWVhc3QtMTowMDAwMDAwMDAwMDA6bXlzcXMuZmlmbyBmNzM1ZjU3Zi1hZjA4LTQ5MWItYTE4OS00YzZiYzFhZGJjZjMgMTcwNjE5ODM4OS42MjgyNjY=",
            "MD5OfBody": "034773867a4b2ca8a0226c09938ac7b5",
            "Body": "{\"Type\": \"Notification\", \"MessageId\": \"caab9f3c-d4ef-4f1e-ba7c-6ba503680f1a\", \"TopicArn\": \"arn:aws:sns:us-east-1:000000000000:mysns.fifo\", \"Message\": \"supermessage\", \"Timestamp\": \"2024-01-25T15:59:45.814Z\", \"UnsubscribeURL\": \"http://localhost.localstack.cloud:4566/?Action=Unsubscribe&SubscriptionArn=arn:aws:sns:us-east-1:000000000000:mysns.fifo:a04a4c8a-1518-429a-9b0d-eb3f9ec5b8e1\", \"SequenceNumber\": \"14656133232500670465\"}",
            "Attributes": {
                "SenderId": "000000000000",
                "SentTimestamp": "1706198385816",
                "MessageGroupId": "mygroupid",
                "MessageDeduplicationId": "superunique2",
                "SequenceNumber": "14656132373507211265",
                "ApproximateReceiveCount": "1",
                "ApproximateFirstReceiveTimestamp": "1706198389628"
            }
        }
    ]
}
```

And I could properly receive both messages one after the other. Could it be a timing issue with message visibility? I believe message visibility is shared across a message group id. 
> When you receive a message with a message group ID, no more messages for the same message group ID are returned unless you delete the message or it becomes visible.

Can you reproduce the issue with something else than the AWS CLI?

**Comment by bentsku** [comment]
Moreover, with a slightly different setup, we can also test the deduplication:

```bash
# same setup of topic, queue et subscription
$ awslocal sns publish --topic-arn "arn:aws:sns:us-east-1:000000000000:mysns.fifo" --message-group-id "mygroupid" --message-deduplication-id "superunique3" --message "supermessage"
{
    "MessageId": "2f25c8b9-b8ca-482e-b099-751c0a082d92",
    "SequenceNumber": "14656139357124034560"
}

$ awslocal sns publish --topic-arn "arn:aws:sns:us-east-1:000000000000:mysns.fifo" --message-group-id "mygroupid" --message-deduplication-id "superunique4" --message "supermessage"
{
    "MessageId": "dbed51f1-e9bc-428b-9274-4c1fb8a71870",
    "SequenceNumber": "14656139357124034561"
}
$ awslocal sqs receive-message --queue-url http://localhost.localstack.cloud:4566/queue/us-east-1/000000000000/mysqs.fifo --attribute-names All
{
    "Messages": [
        {
            "MessageId": "f77e50f4-10f9-47ba-b495-d3c365650fde",
            "ReceiptHandle": "ZGNlYTg1MzktMWE1YS00NGU4LTljMzAtZDc5NDMwZDQzNTMzIGFybjphd3M6c3FzOnVzLWVhc3QtMTowMDAwMDAwMDAwMDA6bXlzcXMuZmlmbyBmNzdlNTBmNC0xMGY5LTQ3YmEtYjQ5NS1kM2MzNjU2NTBmZGUgMTcwNjE5OTA4OC4zNTk1NDE=",
            "MD5OfBody": "c6366341f913176db025690d3fe3c55f",
            "Body": "{\"Type\": \"Notification\", \"MessageId\": \"2f25c8b9-b8ca-482e-b099-751c0a082d92\", \"TopicArn\": \"arn:aws:sns:us-east-1:000000000000:mysns.fifo\", \"Message\": \"supermessage\", \"Timestamp\": \"2024-01-25T16:11:20.711Z\", \"UnsubscribeURL\": \"http://localhost.localstack.cloud:4566/?Action=Unsubscribe&SubscriptionArn=arn:aws:sns:us-east-1:000000000000:mysns.fifo:4560d0f3-48a2-4131-bf78-4cd579fae95f\", \"SequenceNumber\": \"14656139357124034560\"}",
            "Attributes": {
                "SenderId": "000000000000",
                "SentTimestamp": "1706199080741",
                "MessageGroupId": "mygroupid",
                "MessageDeduplicationId": "superunique3",
                "SequenceNumber": "14656138498130575360",
                "ApproximateReceiveCount": "1",
                "ApproximateFirstReceiveTimestamp": "1706199088359"
            }
        }
    ]
}
$ awslocal sqs delete-message --queue-url http://localhost.localstack.cloud:4566/queue/us-east-1/000000000000/mysqs.fifo --receipt-handle ZGNlYTg1MzktMWE1YS00NGU4LTljMzAtZDc5NDMwZDQzNTMzIGFybjphd3M6c3FzOnVzLWVhc3QtMTowMDAwMDAwMDAwMDA6bXlzcXMuZmlmbyBmNzdlNTBmNC0xMGY5LTQ3YmEtYjQ5NS1kM2MzNjU2NTBmZGUgMTcwNjE5OTA4OC4zNTk1NDE=
$ awslocal sqs receive-message --queue-url http://localhost.localstack.cloud:4566/queue/us-east-1/000000000000/mysqs.fifo --attribute-names All
{
    "Messages": [
        {
            "MessageId": "5ebfe928-558a-4158-9b3d-26e896c8dbde",
            "ReceiptHandle": "MWU3MzAxMjktMjMyYS00OTUwLWFlZWUtYTVjZjc4MDYyZmRjIGFybjphd3M6c3FzOnVzLWVhc3QtMTowMDAwMDAwMDAwMDA6bXlzcXMuZmlmbyA1ZWJmZTkyOC01NThhLTQxNTgtOWIzZC0yNmU4OTZjOGRiZGUgMTcwNjE5OTEwMy4yMjM2NTM4",
            "MD5OfBody": "db39deebb3a8694c9ebdc99ee09f6007",
            "Body": "{\"Type\": \"Notification\", \"MessageId\": \"dbed51f1-e9bc-428b-9274-4c1fb8a71870\", \"TopicArn\": \"arn:aws:sns:us-east-1:000000000000:mysns.fifo\", \"Message\": \"supermessage\", \"Timestamp\": \"2024-01-25T16:11:24.082Z\", \"UnsubscribeURL\": \"http://localhost.localstack.cloud:4566/?Action=Unsubscribe&SubscriptionArn=arn:aws:sns:us-east-1:000000000000:mysns.fifo:4560d0f3-48a2-4131-bf78-4cd579fae95f\", \"SequenceNumber\": \"14656139357124034561\"}",
            "Attributes": {
                "SenderId": "000000000000",
                "SentTimestamp": "1706199084084",
                "MessageGroupId": "mygroupid",
                "MessageDeduplicationId": "superunique4",
                "SequenceNumber": "14656138498130575361",
                "ApproximateReceiveCount": "1",
                "ApproximateFirstReceiveTimestamp": "1706199103223"
            }
        }
    ]
}
```

We can see the normal behavior of a FIFO queue, where we can't get the next message as long as this one is not deleted or become visible again. Also, the deduplication seems to not be in effect and we can properly receive both messages.

**Comment by jbauers** [comment]
@bentsku could you have a look here: https://github.com/localstack/localstack/issues/10107#issuecomment-1920869763. Not sure if this is expected behaviour, tried to repro the issue and this is also what we're seeing with our tests failing. Thanks!

**Comment by jamescarter-le** [comment]
I can also replicate this issue as mentioned on the other ticket.

**Comment by bentsku** [comment]
Hello! The underlying issue in SQS should be fixed with #10223. Could you pull the `latest` image with `docker pull localstack/localstack:latest` and see if that fixes your issue? Thanks!

**Comment by jamescarter-le** [comment]
> Hello! The underlying issue in SQS should be fixed with #10223. Could you pull the `latest` image with `docker pull localstack/localstack:latest` and see if that fixes your issue? Thanks!

I belive this has fixed the issue!
It build `3.1.1.dev` with an image hash of `43a367db01b8`, against git hash `7b9fd2c7b`.

Would be happy to bring this back into my dev flow and get away from real SQS for FIFO queues.

**Comment by dandsafe** [comment]
3.1.1.dev works for me. Thank you.
